// hnsw/dist/heap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BinaryHeap = void 0;
class BinaryHeap {
    constructor(compare) {
        this.compare = compare;
        this.items = [];
    }
    get size() {
        return this.items.length;
    }
    peek() {
        return this.items[0];
    }
    push(item) {
        this.items.push(item);
        this.bubbleUp(this.items.length - 1);
    }
    pop() {
        if (this.items.length === 0) {
            return undefined;
        }
        const top = this.items[0];
        const last = this.items.pop();
        if (this.items.length > 0) {
            this.items[0] = last;
            this.bubbleDown(0);
        }
        return top;
    }
    values() {
        return [...this.items];
    }
    bubbleUp(index) {
        let currentIndex = index;
        while (currentIndex > 0) {
            const parentIndex = Math.floor((currentIndex - 1) / 2);
            if (this.compare(this.items[currentIndex], this.items[parentIndex]) <= 0) {
                break;
            }
            this.swap(currentIndex, parentIndex);
            currentIndex = parentIndex;
        }
    }
    bubbleDown(index) {
        let currentIndex = index;
        while (true) {
            const left = currentIndex * 2 + 1;
            const right = currentIndex * 2 + 2;
            let largest = currentIndex;
            if (left < this.items.length && this.compare(this.items[left], this.items[largest]) > 0) {
                largest = left;
            }
            if (right < this.items.length && this.compare(this.items[right], this.items[largest]) > 0) {
                largest = right;
            }
            if (largest === currentIndex) {
                break;
            }
            this.swap(currentIndex, largest);
            currentIndex = largest;
        }
    }
    swap(a, b) {
        const tmp = this.items[a];
        this.items[a] = this.items[b];
        this.items[b] = tmp;
    }
}
exports.BinaryHeap = BinaryHeap;
