// hnsw/dist/bench/dataset.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.streamFvecs = exports.decodeIvecsAsFloatBuffer = exports.readIvecsAsFloat = exports.loadIvecsAsFloatDataset = exports.decodeFvecsBuffer = exports.readFvecs = exports.loadFvecsDataset = exports.generateSyntheticDataset = void 0;
const promises_1 = require("fs/promises");
const fs_1 = require("fs");
const path_1 = require("path");
function generateSyntheticDataset(options) {
    const { name = `synthetic-${options.count}x${options.dimension}`, count, dimension, metric, seed = 42, distribution = 'uniform', } = options;
    const rng = mulberry32(seed);
    const vectors = [];
    for (let i = 0; i < count; i++) {
        const vector = new Float32Array(dimension);
        for (let d = 0; d < dimension; d++) {
            vector[d] = distribution === 'gaussian' ? gaussian(rng) : rng() * 2 - 1;
        }
        vectors.push({ id: i, vector });
    }
    return { name, dimension, metric, vectors };
}
exports.generateSyntheticDataset = generateSyntheticDataset;
function loadFvecsDataset(filePath, metric, options = {}) {
    return __awaiter(this, void 0, void 0, function* () {
        const { limit, offset = 0, asFloat32 = true } = options;
        const name = (0, path_1.basename)(filePath).replace(/\.[^/.]+$/, '');
        const vectors = yield readFvecs(filePath, { limit, offset, asFloat32 });
        const dimension = vectors.length > 0 ? vectors[0].vector.length : 0;
        return { name, dimension, metric, vectors };
    });
}
exports.loadFvecsDataset = loadFvecsDataset;
function readFvecs(filePath, options = {}) {
    return __awaiter(this, void 0, void 0, function* () {
        const { limit, offset = 0, asFloat32 = true } = options;
        const data = yield (0, promises_1.readFile)(filePath);
        return decodeFvecsBuffer(data, { limit, offset, asFloat32 });
    });
}
exports.readFvecs = readFvecs;
function decodeFvecsBuffer(buffer, options = {}) {
    const { limit, offset = 0, asFloat32 = true } = options;
    const vectors = [];
    let cursor = 0;
    let index = 0;
    while (cursor + 4 <= buffer.length) {
        const dim = buffer.readInt32LE(cursor);
        cursor += 4;
        const byteCount = dim * 4;
        if (cursor + byteCount > buffer.length) {
            break;
        }
        const shouldTake = index >= offset && (limit === undefined || vectors.length < limit);
        if (shouldTake) {
            const vector = new Float32Array(dim);
            for (let i = 0; i < dim; i++) {
                vector[i] = buffer.readFloatLE(cursor + i * 4);
            }
            vectors.push({ id: index, vector: asFloat32 ? vector : new Float32Array(vector) });
        }
        cursor += byteCount;
        index++;
    }
    return vectors;
}
exports.decodeFvecsBuffer = decodeFvecsBuffer;
function loadIvecsAsFloatDataset(filePath, metric, options = {}) {
    return __awaiter(this, void 0, void 0, function* () {
        const { limit, offset = 0 } = options;
        const name = (0, path_1.basename)(filePath).replace(/\.[^/.]+$/, '');
        const vectors = yield readIvecsAsFloat(filePath, { limit, offset });
        const dimension = vectors.length > 0 ? vectors[0].vector.length : 0;
        return { name, dimension, metric, vectors };
    });
}
exports.loadIvecsAsFloatDataset = loadIvecsAsFloatDataset;
function readIvecsAsFloat(filePath, options = {}) {
    return __awaiter(this, void 0, void 0, function* () {
        const { limit, offset = 0 } = options;
        const data = yield (0, promises_1.readFile)(filePath);
        return decodeIvecsAsFloatBuffer(data, { limit, offset });
    });
}
exports.readIvecsAsFloat = readIvecsAsFloat;
function decodeIvecsAsFloatBuffer(buffer, options = {}) {
    const { limit, offset = 0 } = options;
    const vectors = [];
    let cursor = 0;
    let index = 0;
    while (cursor + 4 <= buffer.length) {
        const dim = buffer.readInt32LE(cursor);
        cursor += 4;
        const byteCount = dim * 4;
        if (cursor + byteCount > buffer.length) {
            break;
        }
        const shouldTake = index >= offset && (limit === undefined || vectors.length < limit);
        if (shouldTake) {
            const vector = new Float32Array(dim);
            for (let i = 0; i < dim; i++) {
                vector[i] = buffer.readInt32LE(cursor + i * 4);
            }
            vectors.push({ id: index, vector });
        }
        cursor += byteCount;
        index++;
    }
    return vectors;
}
exports.decodeIvecsAsFloatBuffer = decodeIvecsAsFloatBuffer;
function streamFvecs(filePath, onVector, options = {}) {
    var _a, e_1, _b, _c;
    return __awaiter(this, void 0, void 0, function* () {
        const { limit, offset = 0 } = options;
        const stream = (0, fs_1.createReadStream)(filePath);
        let buffer = Buffer.alloc(0);
        let index = 0;
        let emitted = 0;
        try {
            for (var _d = true, stream_1 = __asyncValues(stream), stream_1_1; stream_1_1 = yield stream_1.next(), _a = stream_1_1.done, !_a;) {
                _c = stream_1_1.value;
                _d = false;
                try {
                    const chunk = _c;
                    buffer = Buffer.concat([buffer, chunk]);
                    while (buffer.length >= 4) {
                        const dim = buffer.readInt32LE(0);
                        const byteCount = 4 + dim * 4;
                        if (buffer.length < byteCount) {
                            break;
                        }
                        const recordBuffer = buffer.subarray(0, byteCount);
                        buffer = buffer.subarray(byteCount);
                        const shouldTake = index >= offset && (limit === undefined || emitted < limit);
                        if (shouldTake) {
                            const vector = new Float32Array(dim);
                            for (let i = 0; i < dim; i++) {
                                vector[i] = recordBuffer.readFloatLE(4 + i * 4);
                            }
                            yield onVector({ id: index, vector });
                            emitted++;
                        }
                        index++;
                        if (limit !== undefined && emitted >= limit) {
                            stream.close();
                            return;
                        }
                    }
                }
                finally {
                    _d = true;
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (!_d && !_a && (_b = stream_1.return)) yield _b.call(stream_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
    });
}
exports.streamFvecs = streamFvecs;
function mulberry32(seed) {
    let t = seed >>> 0;
    return () => {
        t += 0x6d2b79f5;
        let r = t;
        r = Math.imul(r ^ (r >>> 15), r | 1);
        r ^= r + Math.imul(r ^ (r >>> 7), r | 61);
        return ((r ^ (r >>> 14)) >>> 0) / 4294967296;
    };
}
function gaussian(rng) {
    let u = 0;
    let v = 0;
    while (u === 0)
        u = rng();
    while (v === 0)
        v = rng();
    return Math.sqrt(-2.0 * Math.log(u)) * Math.cos(2.0 * Math.PI * v);
}
